pisa/stages/flux/pi_mceq_barr_red.py

```diff
@@ -0,0 +1,489 @@
+        gradient_params_mapping["a_f+"] = self.params.barr_a_f_Pi.value.m_as("dimensionless")
+        gradient_params_mapping["g+"] = self.params.barr_g_Pi.value.m_as("dimensionless")
+        gradient_params_mapping["h+"] = self.params.barr_h_Pi.value.m_as("dimensionless")
+        gradient_params_mapping["i+"] = self.params.barr_i_Pi.value.m_as("dimensionless")
+        for k in list(gradient_params_mapping.keys()):
+            gradient_params_mapping[k.replace("+", "-")] = self.antipion_production(
+                gradient_params_mapping[k], pion_ratio
+            )
+
+        # kaons
+        # as the kaon ratio is unknown, K- production is not restricted
+        gradient_params_mapping["w+"] = self.params.barr_w_K.value.m_as("dimensionless")
+        gradient_params_mapping["w-"] = self.params.barr_w_antiK.value.m_as("dimensionless")
+        gradient_params_mapping["x+"] = self.params.barr_x_K.value.m_as("dimensionless")
+        gradient_params_mapping["x-"] = self.params.barr_x_antiK.value.m_as("dimensionless")
+        gradient_params_mapping["y+"] = self.params.barr_y_K.value.m_as("dimensionless")
+        gradient_params_mapping["y-"] = self.params.barr_y_antiK.value.m_as("dimensionless")
+        gradient_params_mapping["z+"] = self.params.barr_z_K.value.m_as("dimensionless")
+        gradient_params_mapping["z-"] = self.params.barr_z_antiK.value.m_as("dimensionless")
+
+        # Populate array Barr param array
+        for (
+            gradient_param_name,
+            gradient_param_idx,
+        ) in self.gradient_param_indices.items():
+            assert gradient_param_name in gradient_params_mapping, (
+                "Gradient parameter '%s' missing from mapping" % gradient_param_name
+            )
+            self.gradient_params[gradient_param_idx] = gradient_params_mapping[
+                gradient_param_name
+            ]
+
+        #
+        # Loop over containers
+        #
+
+        for container in self.data:
+
+            #
+            # Apply the systematics to the flux
+            #
+
+            nubar = container["nubar"]
+            if nubar > 0: flux_key = "nu_flux_nominal"
+            elif nubar < 0: flux_key = "nubar_flux_nominal"
+
+            apply_sys_vectorized(
+                container["true_energy"].get(WHERE),
+                container["true_coszen"].get(WHERE),
+                delta_index,
+                energy_pivot,
+                container[flux_key].get(WHERE),
+                container["gradients"].get(WHERE),
+                self.gradient_params,
+                out=container["nu_flux"].get(WHERE),
+            )
+            container["nu_flux"].mark_changed(WHERE)
+
+            # Check for negative results from spline
+            # TODO - add more spline error/misusage handling
+            # e.g. if events have energy outside spline range throw ERROR
+            negative_mask = container["nu_flux"].get("host") < 0
```

Comment:
this is a very cpu inefficient way to achieve negative values to be clipped. The sum(mask) alone is slow. rather use np.clip() or use if np.any(mask) at least